clarity-ext-scripts/tests/test_rtpcr_analysis_service.py

```diff
@@ -0,0 +1,221 @@
+
+        expected = [
+            {"id": "pos-1", "diagnosis_result": COVID_RESPONSE_POSITIVE},
+            {"id": "pos-2", "diagnosis_result": COVID_RESPONSE_POSITIVE},
+            {"id": "neg-1", "diagnosis_result": COVID_RESPONSE_NEGATIVE},
+            {"id": "neg-2", "diagnosis_result": COVID_RESPONSE_NEGATIVE},
+            {"id": "cov-1", "diagnosis_result": COVID_RESPONSE_POSITIVE},
+            {"id": "cov-2", "diagnosis_result": COVID_RESPONSE_NEGATIVE},
+            {"id": "cov-3", "diagnosis_result": FAILED_BY_TO_HIGH_COVID_VALUE}]
+
+        assert expected == list(result)
+
+    def test_can_analyze_samples_scenario2(self):
+        """
+        Fail on neg control is positive
+        """
+
+        pos_controls = [{"id": "pos-1", "FAM-CT": 25, "HEX-CT": 20},
+                        {"id": "pos-2", "FAM-CT": 20, "HEX-CT": 18}]
+
+        neg_controls = [{"id": "neg-1", "FAM-CT": 20, "HEX-CT": 20},
+                        {"id": "neg-2", "FAM-CT": 0, "HEX-CT": 18}]
+
+        samples = [{"id": "cov-1", "FAM-CT": 20, "HEX-CT": 20},
+                   {"id": "cov-2", "FAM-CT": 0, "HEX-CT": 18},
+                   {"id": "cov-3", "FAM-CT": 40, "HEX-CT": 18}]
+
+        service = RTPCRAnalysisService(
+            covid_reporter_key="FAM", internal_control_reporter_key="HEX")
+
+        result = service.analyze_samples(positive_controls=pos_controls,
+                                         negative_controls=neg_controls,
+                                         samples=samples)
+
+        expected = [
+            {"id": "pos-1", "diagnosis_result": COVID_RESPONSE_POSITIVE},
+            {"id": "pos-2", "diagnosis_result": COVID_RESPONSE_POSITIVE},
+            {"id": "neg-1", "diagnosis_result": COVID_RESPONSE_POSITIVE},
+            {"id": "neg-2", "diagnosis_result": COVID_RESPONSE_NEGATIVE},
+            {"id": "cov-1",
+                "diagnosis_result": FAILED_ENTIRE_PLATE_BY_FAILED_EXTERNAL_CONTROL},
+            {"id": "cov-2",
+                "diagnosis_result": FAILED_ENTIRE_PLATE_BY_FAILED_EXTERNAL_CONTROL},
+            {"id": "cov-3", "diagnosis_result": FAILED_ENTIRE_PLATE_BY_FAILED_EXTERNAL_CONTROL}]
+
+        assert expected == list(result)
+
+    def test_can_analyze_samples_scenario3(self):
+        """
+        Controls are ok. Fail one sample on high HEX value
+        """
+
+        pos_controls = [{"id": "pos-1", "FAM-CT": 25, "HEX-CT": 20},
+                        {"id": "pos-2", "FAM-CT": 20, "HEX-CT": 18}]
+
+        neg_controls = [{"id": "neg-1", "FAM-CT": 0, "HEX-CT": 20},
+                        {"id": "neg-2", "FAM-CT": 0, "HEX-CT": 18}]
+
+        samples = [{"id": "cov-1", "FAM-CT": 20, "HEX-CT": 20},
+                   {"id": "cov-2", "FAM-CT": 0, "HEX-CT": 18},
+                   {"id": "cov-3", "FAM-CT": 0, "HEX-CT": 33}]
+
+        service = RTPCRAnalysisService(
+            covid_reporter_key="FAM", internal_control_reporter_key="HEX")
```

Comment:
Do I understand it correctly the keys will be different between instruments? If so I wonder if we should create subclasses? I.e.

```
SuchAndSuchPCRAnalysisService(RTPCRAnalysisService):
    def __init__(self):
        super(SuchAndSuchPCRAnalysisService, self).__init__("FAM", "HEX")
```

just so it's more readable in the extension.

Reply:
This is a great idea! Lets do that.